### PR TITLE
Callbacks

### PR DESCRIPTION
**Pull request by rachitchaurasia99**
Callbacks
 Product should be always be initialized with default title 'abc' if no title given.
 Discount price should be equal to price unless specified explicitly.
 Send a mail with subject 'Welcome' on users mail after new user is created.
 User should not be destroyed if email is 'admin@depot.com' 
 Should not be able to update user if email is '[admin@depot.com](mailto:admin@depot.com)'

### CODE REVIEW

File: app/mailers/user_mailer.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class UserMailer < ApplicationMailer
+
+  # Subject can be set in your I18n file at config/locales/en.yml
+  # with the following lookup:
+  #
+  #   en.user_mailer.registered.subject
+  #
```
Comment (siddharth-vinsol) [inline app/mailers/user_mailer.rb]: Remove comments

File: app/models/product.rb [hunk 2]
```diff
@@ -14,7 +14,13 @@ class Product < ApplicationRecord
     validates :price, numericality: { greater_than_or_equal_to: 0.01 }
     validates :discount_price, numericality: {less_than_or_equal_to: :price, message: "Discount Price can't be greater than Original Price"}, if: :price 
   end
-
+  
+  before_validation :set_discount_price
+  before_destroy :ensure_not_referenced_by_any_line_item
+  before_validation do |product|
```
Comment (siddharth-vinsol) [inline app/models/product.rb:20]: Dont use blocks for callbacks, use methods

File: app/models/product.rb [hunk 3]
```diff
@@ -14,7 +14,13 @@ class Product < ApplicationRecord
...
+    product.title = 'abc' unless product.title
```
Comment (siddharth-vinsol) [inline app/models/product.rb]: will this work if product title is `'      '`

File: app/models/user.rb [hunk 4]
```diff
@@ -3,13 +3,34 @@ class User < ApplicationRecord
 
   validates :name, presence: true, uniqueness: true
   validates :email, uniqueness: true, format: {with: URI::MailTo::EMAIL_REGEXP}
+  validates :email, uniqueness: true
```
Comment (siddharth-vinsol) [inline app/models/user.rb]: Why this change in callbacks PR

File: app/models/product.rb [hunk 5]
```diff
@@ -14,7 +14,13 @@ class Product < ApplicationRecord
     validates :price, numericality: { greater_than_or_equal_to: 0.01 }
     validates :discount_price, numericality: {less_than_or_equal_to: :price, message: "Discount Price can't be greater than Original Price"}, if: :price 
   end
-
+  
+  before_validation :set_discount_price
```
Comment (siddharth-vinsol) [inline app/models/product.rb:18]: where is this method defined

File: app/models/product.rb [hunk 6]
```diff
@@ -14,7 +14,13 @@ class Product < ApplicationRecord
...
+  before_destroy :ensure_not_referenced_by_any_line_item
```
Comment (siddharth-vinsol) [inline app/models/product.rb:19]: If this method is called as a callback, should this really add errors to `errors` object

File: app/mailers/user_mailer.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+  def registered(user)
+    @greeting = "Welcome"
+
+    mail to: user.email
```
Comment (siddharth-vinsol) [inline app/mailers/user_mailer.rb]: Add subject to mail

File: app/models/user.rb [hunk 8]
```diff
@@ -3,13 +3,34 @@ class User < ApplicationRecord
...
+  
+  def ensure_not_admin
+    if admin?
+      errors.add :base, 'Cannot Update or Delete Admin User'
+      throw :abort
+    end
+  end
+
+  def welcome_mail
+    UserMailer.registered(self).deliver_later
```
Comment (siddharth-vinsol) [inline app/models/user.rb:29]: What if i used `deliver_now` here. What is the difference?
Please mention here

File: app/models/user.rb [hunk 9]
```diff
@@ -3,13 +3,34 @@ class User < ApplicationRecord
 
   validates :name, presence: true, uniqueness: true
   validates :email, uniqueness: true, format: {with: URI::MailTo::EMAIL_REGEXP}
+  validates :email, uniqueness: true
   validates :email, format: { with: URI::MailTo::EMAIL_REGEXP }
 
+  after_destroy :ensure_an_admin_remains
+  after_create_commit :welcome_mail
+  before_update :ensure_not_admin
+  before_destroy :ensure_not_admin
```
Comment (siddharth-vinsol) [inline app/models/user.rb:10]: Make the methods called here as `private`

**Review by siddharth-vinsol — CHANGES_REQUESTED** [review]

File: app/models/user.rb (hunk 8 above)
Comment (rachitchaurasia99) [reply app/models/user.rb:29]: The main difference is that ``` deliver now ``` delivers the mail immediately where as ```deliver_later``` enques the job , so here we can use ```deliver_later``` as we are not doing any authentication operation.

File: app/models/product.rb (hunk 5 above)
Comment (rachitchaurasia99) [inline app/models/product.rb]: Sorry, it missed while pushing the changes, Have corrected it.

File: app/models/product.rb (hunk 6 above)
Comment (rachitchaurasia99) [inline app/models/product.rb:20]: Corrected

File: app/models/product.rb (hunk 2 above)
Comment (rachitchaurasia99) [inline app/models/product.rb]: Done

File: app/models/product.rb (hunk 3 above)
Comment (rachitchaurasia99) [reply app/models/product.rb]: No, Made updates to handle this situation

File: app/models/user.rb (hunk 9 above)
Comment (rachitchaurasia99) [reply app/models/user.rb:10]: Already made them private , at line 15

File: app/mailers/user_mailer.rb (hunk 1 above)
Comment (rachitchaurasia99) [reply app/mailers/user_mailer.rb]: done

File: app/mailers/user_mailer.rb (hunk 7 above)
Comment (rachitchaurasia99) [reply app/mailers/user_mailer.rb]: done

File: app/models/user.rb (hunk 4 above)
Comment (rachitchaurasia99) [reply app/models/user.rb]: fixed

File: app/models/user.rb (hunk 9 above)
Comment (rachitchaurasia99) [reply app/models/user.rb:10]: should i make them private individually?

File: app/models/product.rb [hunk 10]
```diff
@@ -16,14 +16,27 @@ class Product < ApplicationRecord
...
+  before_validation do |product|
+    product.title = 'abc' unless product.title
+  end
+  
   private
 
+  def set_title
+    title = 'abc' unless title || title =~ PRODUCT_TITLE_REGEX
```
Comment (siddharth-vinsol) [inline app/models/product.rb]: this can be simplified

File: app/models/product.rb (hunk 2 above)
Comment (siddharth-vinsol) [reply app/models/product.rb]: You haven't updated this, this is still using block

File: app/models/product.rb [hunk 11]
```diff
@@ -16,14 +16,27 @@ class Product < ApplicationRecord
...
+  end
+
+  def set_discount_price
+    discount_price = price unless discount_price
```
Comment (siddharth-vinsol) [inline app/models/product.rb]: what if someone want `discount_price` as `zero`

File: app/models/product.rb [hunk 12]
```diff
@@ -16,14 +16,27 @@ class Product < ApplicationRecord
...
+  end
+
   # ensure that there are no line items referencing this product
   def ensure_not_referenced_by_any_line_item
     unless line_items.empty?
-      errors.add(:base, 'Line Items present')
-      throw :abort
+      raise 'Product is present in Line Item'
```
Comment (siddharth-vinsol) [inline app/models/product.rb]: what exception will be thrown here and is that the correct exception?
Why not just use `throw :abort` here

File: app/models/user.rb (hunk 9 above)
Comment (siddharth-vinsol) [reply app/models/user.rb:10]: Personal preference to use `private` as block or make methods `private` indiviually
You can go with any one

File: config/initializers/constants.rb [hunk 13]
```diff
@@ -2,3 +2,4 @@
 PERMALINK_REGEX = /\A([\w]+-){2,}[\w]+\Z/
 DESCRIPTION_REGEX = /\A(\w+ ){4,9}\w+ *\z/
 MINIMUM_PRICE = 0.01
+PRODUCT_TITLE_REGEX = /^\s*$/
```
Comment (siddharth-vinsol) [inline config/initializers/constants.rb]: Not required

File: app/mailers/user_mailer.rb
Comment (siddharth-vinsol) [inline app/mailers/user_mailer.rb:1]: Where is the erb file for this

**Review by siddharth-vinsol — CHANGES_REQUESTED** [review]

File: app/models/user.rb (hunk 9 above)
Comment (rachitchaurasia99) [reply app/models/user.rb:10]: ok

File: app/models/product.rb (hunk 10 above)
Comment (rachitchaurasia99) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 11 above)
Comment (rachitchaurasia99) [reply app/models/product.rb]: added numericality for ```discount_price``` also

File: config/initializers/constants.rb (hunk 13 above)
Comment (rachitchaurasia99) [reply config/initializers/constants.rb]: fixed

File: app/mailers/user_mailer.rb
Comment (rachitchaurasia99) [reply app/mailers/user_mailer.rb:1]: done

File: app/models/product.rb (hunk 12 above)
Comment (rachitchaurasia99) [reply app/models/product.rb]: done

**Review by siddharth-vinsol — APPROVED** [review]
